core/ast_node.js

```diff
@@ -0,0 +1,656 @@
+  if (!newAstNode && curIdx - 1 >= 0) {
+    newAstNode = Blockly.ASTNode.createInputNode(inputs[curIdx - 1]);
+  }
+  return newAstNode;
+};
+
+/**
+ * Navigate between stacks of blocks on the workspace.
+ * @param {?Boolean} forward True to go forward. False to go backwards.
+ * @return {?Blockly.BlockSvg} The first block of the next stack.
+ * @private
+ */
+Blockly.ASTNode.prototype.navigateBetweenStacks_ = function(forward) {
+  var curLocation = this.getLocation();
+  if (!(curLocation instanceof Blockly.Block)) {
+    curLocation = curLocation.sourceBlock_;
+  }
+  if (!curLocation) {
+    return null;
+  }
+  var curRoot = curLocation.getRootBlock();
+  var topBlocks = curRoot.workspace.getTopBlocks();
+  for (var i = 0; i < topBlocks.length; i++) {
+    var topBlock = topBlocks[i];
+    if (curRoot.id == topBlock.id) {
+      var offset = forward ? 1 : -1;
+      var resultIndex = i + offset;
+      if (resultIndex == -1) {
+        resultIndex = topBlocks.length - 1;
+      } else if (resultIndex == topBlocks.length) {
+        resultIndex = 0;
+      }
+      return topBlocks[resultIndex];
+    }
+  }
+  throw Error('Couldn\'t find ' + (forward ? 'next' : 'previous') +
+      ' stack?!?!?!');
+};
+
+/**
+ * Create an ast node from the top connection on the block.
+ * @param {Blockly.Block} block The block that we want to find the top
+ *     connection on.
+ * @return {?Blockly.ASTNode} The astnode containing the top connection.
+ * @private
+ */
+Blockly.ASTNode.prototype.findTopASTConnection_ = function(block) {
+  var previousConnection = block.previousConnection;
+  var outputConnection = block.outputConnection;
+  var newAstNode;
+  if (previousConnection) {
+    newAstNode = Blockly.ASTNode.createConnectionNode(previousConnection);
+  } else if (outputConnection) {
+    newAstNode = Blockly.ASTNode.createConnectionNode(outputConnection);
+  }
+  return newAstNode;
+};
+
+
+/**
+ * Given a location in a stack of blocks find the next out connection. If the
+ * location is nested the next out location should be the connected input.
+ * @param {!Blockly.Block} block The source block for the current location.
+ * @return {?Blockly.ASTNode} The ast node holding thenext out connection or
+ *     block.
+ * @private
+ */
+Blockly.ASTNode.prototype.findOutLocationForBlock_ = function(block) {
+  var newLocation, newAstNode;
+
+  newLocation = block.previousConnection.targetConnection;
+  newAstNode = Blockly.ASTNode.createConnectionNode(newLocation);
+  if (!newLocation) {
+    newLocation = block.previousConnection;
+    newAstNode = Blockly.ASTNode.createStackNode(block);
+  }
+  return newAstNode;
+};
+
+/**
+ * Find the element to the right of the current element in the ast.
+ * @return {?Blockly.ASTNode} An ast node that wraps the next field, connection,
+ *     block, or workspace. Return null if there is no value to the right.
+ */
+Blockly.ASTNode.prototype.next = function() {
+  var newAstNode = null;
+
+  switch (this.type_) {
```

Comment:
For each case in next/prev/in/out please add a comment about what you expect the returned node in the tree to be. For example, "If on a field next should return the next editable field or input connection on the block. It will not move between blocks and it will return null if this is the last valid field or input on the block."